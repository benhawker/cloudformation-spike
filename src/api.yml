swagger: "2.0"
schemes:
  - https
info:
  title: Product API
  version: 0.0.1
  description: |
    Example API Gateway implementation
x-amazon-apigateway-request-validators:
  $ref: ./gateway/validators.yml
x-amazon-apigateway-gateway-responses:
  $ref: ./gateway/responses.yml
definitions:
  $ref: ./definitions.yml
paths:
  /health:
    get:
      $ref: ./operations/getHealth.yml
  /products:
    get:
      $ref: ./operations/getProducts/getProducts.yml
    post:
      $ref: ./operations/postProduct/postProduct.yml
  /products/{id}:
    get:
      $ref: ./operations/getProduct/getProduct.yml