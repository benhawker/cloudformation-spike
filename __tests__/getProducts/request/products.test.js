import api from 'api-gateway-template-tester';
import * as helpers from '../../__utils__/testHelpers.js';

// Load the request template that we are testing.
// I.e. template that we use to build POST request to DynamoDB.
const template = global.loadVelocityTemplate(
  'src/operations/getProducts/request.yml', 'application/json');


describe('Validating request templates', () => {
  let test;
  let interpolatedTemplate;
  let parsedTemplate;

  const itemsArray = [helpers.createProduct(1), helpers.createProduct(2), helpers.createProduct(3)];

  beforeAll((done) => {
    helpers.ddb.createTable(helpers.tableParams, () => {
      helpers.ddb.putItem(itemsArray[0], () => {
        helpers.ddb.putItem(itemsArray[1], () => {
          helpers.ddb.putItem(itemsArray[2], done);
        });
      });
    });

    test = api().stageVariables({ ProductsTable: helpers.tableName });
    interpolatedTemplate = test.render(template);
    parsedTemplate = JSON.parse(interpolatedTemplate);
  });

  afterAll((done) => {
    helpers.ddb.deleteTable({ TableName: helpers.tableName }, done);
  });

  describe('returns the correct response', () => {
    it('Produced valid JSON', (done) => {
      helpers.ddb.scan(parsedTemplate, (err, data) => {
        expect(data).toMatchSnapshot();
        done();
      });
    });

    it('interpolates and parses the template correctly', () => {
      const expectedTemplate = { TableName: helpers.tableName };
      expect(JSON.parse(interpolatedTemplate)).toEqual(expectedTemplate);
    });

    it('is valid JSON', () => {
      expect(helpers.isValidJson(template)).toEqual(true);
    });
  });
});
