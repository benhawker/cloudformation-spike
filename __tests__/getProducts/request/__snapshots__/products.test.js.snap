// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validating request templates returns the correct response Produced valid JSON 1`] = `
Object {
  "Count": 3,
  "Items": Array [
    Object {
      "description": Object {
        "S": "Another really great product description",
      },
      "id": Object {
        "S": "1",
      },
      "price": Object {
        "N": "100",
      },
      "title": Object {
        "S": "Product Name 1",
      },
    },
    Object {
      "description": Object {
        "S": "Another really great product description",
      },
      "id": Object {
        "S": "3",
      },
      "price": Object {
        "N": "100",
      },
      "title": Object {
        "S": "Product Name 3",
      },
    },
    Object {
      "description": Object {
        "S": "Another really great product description",
      },
      "id": Object {
        "S": "2",
      },
      "price": Object {
        "N": "100",
      },
      "title": Object {
        "S": "Product Name 2",
      },
    },
  ],
  "ScannedCount": 3,
}
`;
